### ISSUE TITLE
Wesnoth creates D:\a\msys64\var\cache\fontconfig with two files inside

### ISSUE THREAD

**Issue by Toranks** [wesnoth/wesnoth#7089]
### Game and System Information

The Battle for Wesnoth version 1.17.9 x86_64
Running on Microsoft Windows 7 Service Pack 1 (6.1.7601) x86_64
Distribution channel: Steam

### Description of the bug

While the game is active it won't let me delete the files, but even if I delete the entire path, when I start the game it is created again.

### Steps to reproduce the behavior

1. Start the game
2. Go to the main folder of the drive (X:) when the game is installed
3. X:\a\msys64\var\cache\fontconfig appears with two cache files inside

### Expected behavior

Store cache files on
Cache dir:       C:\Users\USER\Documents\My Games\Wesnoth1.17\cache

### Additional context

Possibly only happens on Windows 7

**Comment by Wedge009** [comment]
I confirm this using the 1.17.9 Windows installer package from SourceForge. But only on Windows 7, I couldn't find a similar directory for Windows 10 - and I have no idea why that would be the case. I'm not aware of the details of how Wesnoth is cross-compiled for Windows releases.

For my Windows 7 set-up I have C-drive for boot (including Windows installation), D-drive for data, X-drive as a RAM-disk and Z-drive for back-up. The fontconfig cache went to D-drive for me.

**Comment by Wedge009** [comment]
Revisited this while investigating something else. Confirming this was introduced at some point between 1.17.4 and 1.17.5.

**Comment by Pentarctagon** [comment]
It's possible I suppose?

**Comment by Wedge009** [comment]
I'll check if 1.16 starts having the same problem if/when it gets back-ported then.

**Comment by Pentarctagon** [comment]
@loonycyborg Could this potentially be solved by setting a `<cachedir>` in https://github.com/wesnoth/wesnoth/blob/master/fonts/fonts.conf?

**Comment by knyghtmare** [comment]
Confirmed on Windows 10.

**Comment by Wedge009** [comment]
What's your drive layout? Maybe it happens only if a D-drive is available.

**Comment by knyghtmare** [comment]
I have a main C Drive and a second D Drive...and that's it.

![image](https://user-images.githubusercontent.com/5283677/232423361-91271743-0893-4592-939b-7fe26b5516a6.png)

**Comment by Wedge009** [comment]
So presumably there's a hard-coded desire to write to a D-drive in fontconfig somewhere. If there isn't a D-drive, it just doesn't have a cache.

**Comment by loonycyborg** [comment]
dir `a` seems like it's coming from `git diff`. Maybe it's something about default settings in msys2 build of fontconfig.

**Comment by loonycyborg** [comment]
> @loonycyborg Could this potentially be solved by setting a `<cachedir>` in https://github.com/wesnoth/wesnoth/blob/master/fonts/fonts.conf?

It's worth a try at very least.

**Comment by Pentarctagon** [comment]
@Toranks @Wedge009 or @knyghtmare can you try adding `<cachedir>C:/path/to/somewhere/else</cachedir>` to line 33 of https://github.com/wesnoth/wesnoth/blob/master/fonts/fonts.conf and see if it changes where these files are written?

**Comment by knyghtmare** [comment]
>  can you try adding `<cachedir>C:/path/to/somewhere/else</cachedir>` to line 33 of https://github.com/wesnoth/wesnoth/blob/master/fonts/fonts.conf and see if it changes where these files are written?

I made the adjustment and launched the client. I dont see it in D Drive anymore.

**Comment by Pentarctagon** [comment]
So then presumably `<cachedir>~/wesnoth-fontconfig</cachedir>` would put it in `C:/Users/USER/wesnoth-fontconfig/`, though that would also affect non-Windows platforms.

**Comment by loonycyborg** [comment]
There might be way to override it with env variable too.

**Comment by Toranks** [comment]
The change works for 1.17.15 and 1.16.9+dev from Pentarctagon build he sent in #development

**Comment by Pentarctagon** [comment]
Yeah an environment variable would be best, but I don't see one listed for this on https://www.freedesktop.org/software/fontconfig/fontconfig-user.html (scroll to the very end).

**Comment by Pentarctagon** [comment]
Seems to be an unresolved issue with GIMP as well: https://gitlab.gnome.org/GNOME/gimp/-/issues/5940

**Comment by Wedge009** [comment]
I thought the change in #7547 was included in 1.17.16 release. I just tried it (downloaded from sourceforge) and still have the cache files being generated. Or is the sourceforge installer not correct? eg I see the Wesnoth executables still showing version 1.17.15, which may be a recurrence of #7441.

**Comment by Pentarctagon** [comment]
Where is it showing 1.17.15?

**Comment by Wedge009** [comment]
In the Windows Explorer file properties, as per #7441. Within the game itself, it still claims 1.17.16.

**Comment by Pentarctagon** [comment]
Hmm...

Can you add `PLAIN_LOG << font_file_contents;` on line 171 of `src/font/font_config.cpp` and post the output here?

**Comment by Toranks** [comment]
It also happens to me that the executable says it has version 1.17.15 even though the game is 16. But the cache file is created correctly in the correct folder now, I have checked it explicitly

**Comment by Wedge009** [comment]
@Pentarctagon I finally recompiled master @ 56c2b1840104a9d66aebff9a0f14ac7b7a07beb7 and when I run it I don't see any `a` directory in my D-drive. But I'm not sure if that's a sign that this issue is resolved because I don't recall encountering this issue from a self-compiled executable.

@Toranks What is the 'correct' directory for the fontconfig cache? Just so I know what to look for.

**Comment by Pentarctagon** [comment]
It should be the regular wesnoth cache directory.

**Comment by Toranks** [comment]
![imagen](https://user-images.githubusercontent.com/40789364/235285414-d0f6f5e8-bd4e-46f1-9e78-4b4720ba2013.png)
The file that I point to is the one that previously appeared in the other folder. I can't find the other file, but I think it's not created anymore

**Comment by Wedge009** [comment]
Hmm, something odd is going on because after clearing my `Wesnoth1.17` cache directory, I only see the usual `cache-v1.17.16+dev-[hash].gz` files, nothing that looks like it should be from fontconfig.

What makes it more confusing is when I log the `font_file_contents` as per Pent's instructions, I get this from the log:
`<cachedir>D:\USER\Documents\My Games\Wesnoth1.17\cache</cachedir>`

That looks correct to me because when I debug `manager::manager()`, what's rendered as `USER` is actually my Windows user name. So the configuration looks right... I'm just not getting anything written there.

But all of that doesn't explain why 1.17.16 release build is still creating a `D:\a\...` cache location while it doesn't appear to be the case for Toranks.

**Comment by Pentarctagon** [comment]
@Toranks Where did you download Wesnoth from?

**Comment by Toranks** [comment]
Steam

**Comment by Wedge009** [comment]
I just tried master on another Windows machine which only has a C-drive (and thus never exhibited this issue), and curiously it doesn't appear to generate the fontconfig cache in the Wesnoth cache directory either. When I debug `manager::manager()`, it's also showing the `<cachedir>` addition at the right directory - but nothing (besides the usual `gz` files) gets generated.

**Comment by Wedge009** [comment]
I'm re-opening this. I just tested 1.17.17 release (downloaded from sourceforge.net) and it's still producing the `D:\a\...` files and directories. I'm not sure how my configuration differs from Toranks, given the issue seems to be resolved for him.

**Comment by knyghtmare** [comment]
> I'm re-opening this. I just tested 1.17.17 release (downloaded from sourceforge.net) and it's still producing the `D:\a\...` files and directories. I'm not sure how my configuration differs from Toranks, given the issue seems to be resolved for him.

Confirmed (again) on Windows 10

**Comment by Pentarctagon** [comment]
Alright, well, if that doesn't work then I think we're out of options aside from waiting for upstream to fix the bug.

**Comment by Wedge009** [comment]
It appears this only shows itself on drives besides `C:\` - I just noticed a Windows 10 laptop with a `D:\` also has this problem (dated March 2024), re-confirming aforementioned reports.

**Comment by Psimage** [comment]
Confirming the issue is present on:
* Windows 11
* Steam version of the game v1.18.5
* I have C,D,E drives. SteamLibrary is on D, the "a" directory also on D
* Uninstalling the game does not uninstall "a" directory

**Comment by Morridin** [comment]
I presume the behaviour of Wesnoth not creating the `a` folder in `C:\` root directory is because of privilege issues. I just disallowed users to write in my `D:\` drive's root folder and - behold! - the folder didn't show up on game start any more.

I faced that issue on Windows 11, with Wesnoth 1.18.5 installed via Steam on `D:` drive.

**Comment by soliton-** [comment]
Wesnoth does not intend to create the `a` directory at all. It's a quirk of fontconfig on windows.

**Comment by Wedge009** [comment]
Curious, I asked a LLM about this:

> **The core problem: fontconfig's cache directory is baked in at compile time.**
When fontconfig (as distributed through MSYS2/MinGW) is built, the default cache directory path gets hardcoded into the binary based on where it was compiled, not where it's installed or run. If fontconfig is used with a fonts.conf that does not define any cache directories, fontconfig falls back to that hardcoded path — the build machine's MSYS2 prefix — rather than using a sensible runtime default. [GitHub](https://github.com/msys2/MINGW-packages/issues/2954)
So the D:\a\msys64\... path you're seeing is almost certainly the path that was present on the GitHub Actions runner (or similar CI machine) where the MSYS2/MinGW fontconfig package was compiled. The `D:\a\` prefix is a telltale sign — that's the default workspace root on GitHub Actions Windows runners. The binary then gets distributed to end users, who end up with fontconfig trying to write caches to that baked-in CI path on their own machines.

We have already tried the explicit cache directory it suggested, it sounds like this is a non-intrusive but otherwise annoying bug that really needs to be fixed in MSYS2.
